### PR TITLE
SWATCH-261 Schedule and process hourly tallies by org_id instead of a…

### PR DESCRIPTION
**Pull request by kartikshahc**
…ccount

### CODE REVIEW

**Comment by kartikshahc** [comment]
https://issues.redhat.com/browse/SWATCH-261

**1. Description:**
- We have producer code CaptureSnapshotsTaskManager which is being called by a cron job and JMX endpoint hourly tally. It queues the org stream based on the batch size. The queue flushes the stream with orgId as Arg.
- The consumer TallyTaskFactory reads the kafka queue for orgId as Arg. Later snapshot controller is called to produce hourly snapshot for org.

**Add test data:**
```INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, account_number, subscription_number, billing_provider, billing_account_id) VALUES ('MW01484', 'org123', '12351253', 1, '2012-05-13 16:32:00.000000 +00:00', '2032-05-13 16:32:00.000000 +00:00', '1', '', '1235153', 'red hat', '456');
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, account_number, subscription_number, billing_provider, billing_account_id) VALUES ('MW01882', 'org123', '12351255', 1, '2012-05-13 16:32:00.000000 +00:00', '2032-05-13 16:32:00.000000 +00:00', '1', '', '1235153', 'red hat', '456');
INSERT INTO public.account_config (account_number, opt_in_type, created, updated, org_id) VALUES ('account123', 'JMX', '2022-11-02 17:07:12.062506 +00:00', '2022-11-02 17:07:12.062506 +00:00', 'org123');
INSERT INTO public.billable_usage_remittance (account_number, org_id, product_id, metric_id, accumulation_period, sla, usage, billing_provider, billing_account_id, remitted_value, remittance_date, version) VALUES ('account123', 'org123', 'rhosak', 'Cores', '2022-08', 'Premium', 'Production', 'red hat', 'dummyId', 9, '2022-11-02 19:58:49.424626 +00:00', 0);
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('09443c41-a773-4e22-970b-07256b0bf43d', 'account123', '2022-08-25 20:00:00.000000 +00:00', '{"sla": "Premium", "role": "osd", "org_id": "org123", "event_id": "09443c41-a773-4e22-970b-07256b0bf43d", "timestamp": "2021-10-18T20:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:4cpu_hour", "expiration": "2021-10-19T20:00:00Z", "instance_id": "aeb1f778-af79-4708-8401-35a5a9e8e877", "display_name": "aeb1f778-af79-4708-8401-35a5a9e8e877", "event_source": "prometheus", "measurements": [{"uom": "Cores", "value": 8.4}], "service_type": "OpenShift Cluster", "account_number": "account123", "billing_provider": "red hat", "billing_account_id": "dummyId"}', 'snapshot_redhat.com:openshift_dedicated:4cpu_hour', 'prometheus', 'aeb1f778-af79-4708-8401-35a5a9e8e877', 'org123');
```

**2A. Test Steps with JMX tallyAllAccountsByHourly endpoint.**
- Run USER_USE_STUB=true SUBSCRIPTION_USE_STUB=true CG_USE_STUB=true CLOUDIGRADE_ENABLED=false DEV_MODE=true ./gradlew clean :bootRun
- Execute JMX endpoint tallyAllAccountsByHourly()

**2A. Results**
2022-11-02 20:46:35,291 [thread=http-nio-9000-exec-9] [INFO ] [org.candlepin.subscriptions.tally.job.CaptureSnapshotsTaskManager] - Queuing hourly snapshot production for orgId org123 between 2022-08-03T10:00:00Z and 2022-10-27T10:00:00Z
2022-11-02 20:46:35,291 [thread=http-nio-9000-exec-9] [INFO ] [org.candlepin.subscriptions.task.queue.kafka.KafkaTaskQueue] - Queuing task: TaskDescriptor[groupId: platform.rhsm-subscriptions.tasks, taskType: UPDATE_HOURLY_SNAPSHOTS, args: [startDateTime: [2022-08-03T10:00:00Z], endDateTime: [2022-10-27T10:00:00Z], orgId: [org123]]]

2022-11-02 20:46:38,285 [thread=rhsm-subscriptions-task-processor-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.SnapshotSummaryProducer] - Produced 32 TallySummary messages
2022-11-02 20:46:38,285 [thread=rhsm-subscriptions-task-processor-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.TallySnapshotController] - Finished producing hourly snapshots for account account123 with orgId org123

image

**2B. Test Steps with JMX tallyOrgByHourly or tallyAccountByHourly endpoint.**
- Run USER_USE_STUB=true SUBSCRIPTION_USE_STUB=true CG_USE_STUB=true CLOUDIGRADE_ENABLED=false DEV_MODE=true ./gradlew clean :bootRun
- Execute JMX endpoint tallyOrgByHourly()

**2B. Results**
2022-11-02 20:46:35,291 [thread=http-nio-9000-exec-9] [INFO ] [org.candlepin.subscriptions.tally.job.CaptureSnapshotsTaskManager] - Queuing hourly snapshot production for orgId org123 between 2022-08-03T10:00:00Z and 2022-10-27T10:00:00Z
2022-11-02 20:46:35,291 [thread=http-nio-9000-exec-9] [INFO ] [org.candlepin.subscriptions.task.queue.kafka.KafkaTaskQueue] - Queuing task: TaskDescriptor[groupId: platform.rhsm-subscriptions.tasks, taskType: UPDATE_HOURLY_SNAPSHOTS, args: [startDateTime: [2022-08-03T10:00:00Z], endDateTime: [2022-10-27T10:00:00Z], orgId: [org123]]]

2022-11-02 20:46:38,285 [thread=rhsm-subscriptions-task-processor-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.SnapshotSummaryProducer] - Produced 32 TallySummary messages
2022-11-02 20:46:38,285 [thread=rhsm-subscriptions-task-processor-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.TallySnapshotController] - Finished producing hourly snapshots for account account123 with orgId org123

- Also in both cases you will see an entry in account_services table
e.g.: ```account123,OpenShift Cluster,org123```

**Review by mstead — CHANGES_REQUESTED** [review]
Couple of small items to address, but the code looks good and seems to function properly.

**Review by mstead — APPROVED** [review]

File: src/main/java/org/candlepin/subscriptions/tally/job/CaptureHourlySnapshotsJob.java [hunk 1]
```diff
@@ -41,7 +45,8 @@ public CaptureHourlySnapshotsJob(CaptureSnapshotsTaskManager taskManager) {
   @Scheduled(cron = "${rhsm-subscriptions.jobs.capture-hourly-snapshot-schedule}")
   public void run() {
     try {
-      tasks.updateHourlySnapshotsForAllAccounts(Optional.empty());
+      log.info("Hourly all org tally snapshot triggered by cron job");
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureHourlySnapshotsJob.java]: Done

File: src/main/java/org/candlepin/subscriptions/jmx/TallyJmxBean.java [hunk 2]
```diff
@@ -104,28 +105,54 @@ public void tallyAccountByHourly(
...
+          "Beginning of the range of time the tally should include. "
+              + "Should be top of the hour and expected to be in ISO 8601 format (e.g. 2020-08-02T14:00Z).")
+  @ManagedOperationParameter(
+      name = "endDateTime",
+      description =
+          "Ending of the range of time the tally should include. "
+              + "Should be top of the hour and expected to be in ISO 8601 format (e.g. 2020-08-02T14:00Z).")
+  @ParameterDuration("@jmxProperties.tallyBean.hourlyTallyDurationLimitDays")
+  public void tallyOrgByHourly(
+      String orgId, @NotNull String startDateTime, @NotNull String endDateTime) {
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/jmx/TallyJmxBean.java]: Done

**Pull request merged**